File: app/util/WeightedPriceCalculator.java
package util;

import java.util.Calendar;
import java.util.Collections;
import java.util.Map;

import model.SpotPriceInfo;
import play.Logger;

import com.amazonaws.services.ec2.model.DescribeSpotPriceHistoryRequest;
import com.amazonaws.services.ec2.model.DescribeSpotPriceHistoryResult;
import com.amazonaws.services.ec2.model.InstanceType;
import com.amazonaws.services.ec2.model.SpotPrice;

public class WeightedPriceCalculator implements BaseAmazonClient
{
    public void addWeightedPrice(InstanceType instanceType, String availabilityZone, Map<InstanceType,SpotPriceInfo> lowestWeightedPrices)
    {
        double currentPrice = getCurrentPrice(instanceType, availabilityZone);
        double lastDayAverage = getLastDayAverage(instanceType, availabilityZone);
        double threeMonthAverage = getThreeMonthAverage(instanceType, availabilityZone);
        double weightedPrice = currentPrice * .25 + lastDayAverage * .25 + threeMonthAverage * .5;
        weightedPrice = Math.floor(weightedPrice * 10000) / 10000; //Round to four decimals
        synchronized (lowestWeightedPrices)
        {
            if (!lowestWeightedPrices.containsKey(instanceType))
                lowestWeightedPrices.put(instanceType, new SpotPriceInfo(instanceType, availabilityZone, weightedPrice));
            else
                if (lowestWeightedPrices.get(instanceType).price > weightedPrice)
                    lowestWeightedPrices.get(instanceType).setSpotPrice(new String(availabilityZone), weightedPrice);
        }
    }

    private double getThreeMonthAverage(InstanceType instanceType,
            String availabilityZone)
    {
        return getAveragePrice(instanceType, availabilityZone, Calendar.MONTH, 3);
    }

    private double getLastDayAverage(InstanceType instanceType,
            String availabilityZone)
    {
        return getAveragePrice(instanceType, availabilityZone, Calendar.HOUR, 24);
    }

    private double getCurrentPrice(InstanceType instanceType,
            String availabilityZone)
    {
        return getAveragePrice(instanceType, availabilityZone, Calendar.MINUTE, 1);
    }
    
    private double getAveragePrice(InstanceType instanceType, String availabilityZone, int calendarField, int period)
    {
        DescribeSpotPriceHistoryRequest priceHistoryRequest = new DescribeSpotPriceHistoryRequest();
        priceHistoryRequest.setInstanceTypes(Collections.singleton(instanceType.toString()));
        Calendar calendar = Calendar.getInstance();
        priceHistoryRequest.setEndTime(calendar.getTime());
        calendar.add(calendarField, -period);
        priceHistoryRequest.setStartTime(calendar.getTime());
        priceHistoryRequest.setProductDescriptions(Collections.singleton("Linux/UNIX"));
        priceHistoryRequest.setAvailabilityZone(availabilityZone);
        DescribeSpotPriceHistoryResult spotPriceHistory;
        double accumulator = 0.0;
        int numberOfResults = 0;
        do
        {
            spotPriceHistory = ec2ClientAsync.describeSpotPriceHistory(priceHistoryRequest);
            numberOfResults += spotPriceHistory.getSpotPriceHistory().size();
            for(SpotPrice spotPrice : spotPriceHistory.getSpotPriceHistory())
                accumulator += Double.valueOf(spotPrice.getSpotPrice());
            priceHistoryRequest.setNextToken(spotPriceHistory.getNextToken());
        }
        while(!"".equals(spotPriceHistory.getNextToken()));
        if(numberOfResults == 0)
            return Double.MAX_VALUE;
        return accumulator / numberOfResults;
    }
}

File: app/util/WeightedPriceCalculator.scala
package util

import com.amazonaws.services.ec2.model.InstanceType
import scala.collection.Map
import model.SpotPriceInfo
import java.util.Calendar
import com.amazonaws.services.ec2.model.DescribeSpotPriceHistoryRequest
import java.util.Collections
import com.amazonaws.services.ec2.model.DescribeSpotPriceHistoryResult
import com.amazonaws.services.ec2.model.SpotPrice
import scala.collection.JavaConversions._
import scala.concurrent.Future
import scala.concurrent.Await
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext
import play.libs.Akka
import scala.concurrent.ExecutionContext.Implicits.global
import play.Logger
import scala.annotation.tailrec

object WeightedPriceCalculator extends AmazonClient {

    private final val LinuxProductDescription: String = "Linux/UNIX"

	def getWeightedPrice(implicit instanceType: InstanceType, availabilityZone: String): Double = {
	    
		val currentPriceFut = Future {
		    getCurrentPrice(instanceType, availabilityZone)
		}
		val lastDayAverageFut = Future {
		    getLastDayAverage(instanceType, availabilityZone)
		}
		val threeMonthAverageFut = Future {
		    getThreeMonthAverage(instanceType, availabilityZone)
		}
		
		Await.result (
		for {
		    currentPrice <- currentPriceFut
		    lastDayAverage <- lastDayAverageFut
		    threeMonthAverage <- threeMonthAverageFut
		}
		yield Math.floor(currentPrice.getOrElse(lastDayAverage.get) * .25 + lastDayAverage.get * .25 + threeMonthAverage.get * .5 * 10000) / 10000 //Round to four decimals
		, Duration.Inf
		)
	}

	def getCurrentPrice(instanceType: InstanceType, availabilityZone: String) = getAveragePrice(instanceType, availabilityZone, Calendar.MINUTE, 1)

	private[this] def getLastDayAverage(instanceType: InstanceType, availabilityZone: String) = getAveragePrice(instanceType, availabilityZone, Calendar.HOUR, 24)

    private[this] def getThreeMonthAverage(instanceType: InstanceType, availabilityZone: String) = getAveragePrice(instanceType, availabilityZone, Calendar.MONTH, 3)

    private[this] def getAveragePrice(instanceType: InstanceType, availabilityZone: String, calendarField: Int, period: Int): Option[Double] = {
    	val priceHistoryRequest: DescribeSpotPriceHistoryRequest = new DescribeSpotPriceHistoryRequest
    	priceHistoryRequest.setInstanceTypes(Collections.singleton(instanceType.toString))
    	val calendar: Calendar = Calendar.getInstance
    	priceHistoryRequest.setEndTime(calendar.getTime)
    	calendar.add(calendarField, -period)
    	priceHistoryRequest.setStartTime(calendar.getTime)
    	priceHistoryRequest.setProductDescriptions(Collections.singleton(LinuxProductDescription))
    	priceHistoryRequest.setAvailabilityZone(availabilityZone)
    	getAverageRecursive(0.0, 0, priceHistoryRequest)
    }

    @tailrec private[this] def getAverageRecursive(accumulatedSum: Double, numberOfResults: Int, priceHistoryRequest: DescribeSpotPriceHistoryRequest): Option[Double] = {
    	val spotPriceHistory: DescribeSpotPriceHistoryResult = ec2ClientAsync.describeSpotPriceHistory(priceHistoryRequest)
    	val sumOfThis: Double = spotPriceHistory.getSpotPriceHistory.map{spotPrice: SpotPrice => spotPrice.getSpotPrice.toDouble}.sum
    	val accumulatedSumThis: Double = accumulatedSum + sumOfThis
    	val numberOfResultsThis: Int = numberOfResults + spotPriceHistory.getSpotPriceHistory.size
    	if(spotPriceHistory.getNextToken.isEmpty)
    	{
    		if (numberOfResultsThis == 0) None else Some(accumulatedSumThis / numberOfResultsThis)
    	}
    	else
    	{
    		priceHistoryRequest.setNextToken(spotPriceHistory.getNextToken)
    		getAverageRecursive(accumulatedSumThis, numberOfResultsThis, priceHistoryRequest)
    	}
    }
}